### PR TITLE
log warning when filter conversion/bind expression fail

### PR DESCRIPTION
**Pull request by huaxingao**
Log a warning when filter conversion/bind expression fail to address this comment https://github.com/apache/iceberg/pull/5227#discussion_r917098256

### CODE REVIEW

**Comment by huaxingao** [comment]
cc @szehon-ho @bryanck

**Review by bryanck — APPROVED** [review]

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java [hunk 1]
```diff
@@ -105,6 +108,7 @@ public Filter[] pushFilters(Filter[] filters) {
         expr = SparkFilters.convert(filter);
       } catch (IllegalArgumentException e) {
         // converting to Iceberg Expression failed, so this expression cannot be pushed down
+        LOG.warn("converting to Iceberg Expression failed, so this expression cannot be pushed down.", e);
```
Comment (szehon-ho) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:111]: Sorry to be nitty, but capitlization does not seem right.  What do think like 'Failed to convert filter to Iceberg expression, skipping push down").  

And I wonder, does it help to add the filter to the message?

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java [hunk 2]
```diff
@@ -114,6 +118,7 @@ public Filter[] pushFilters(Filter[] filters) {
           pushed.add(filter);
         } catch (ValidationException e) {
           // binding to the table schema failed, so this expression cannot be pushed down
+          LOG.warn("binding to the table schema failed, so this expression cannot be pushed down.", e);
```
Comment (szehon-ho) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:121]: What do you think, "Failed to bind expression to table schema, skipping push down"?

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java (hunk 1 above)
Comment (kbendick) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:111]: +1 to adding the filter to the message.

So here, I would think maybe LOG.warn(" 'Failed to convert filter to Iceberg expression, skipping push down for this expression: {}", filter, e);'

**Review by kbendick — APPROVED** [review]
LGTM. Thanks @huaxingao!

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java (hunk 1 above)
Comment (huaxingao) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java]: Sounds good! Changed.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java (hunk 2 above)
Comment (huaxingao) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java]: Changed. Thanks!

**Comment by szehon-ho** [comment]
Merged, thanks @huaxingao for change, @bryanck @kbendick for review!

**Comment by huaxingao** [comment]
Thank you very much! @szehon-ho @bryanck @kbendick

**Comment by szehon-ho** [comment]
HI @huaxingao @kbendick @bryanck , I see this log now in some instances, for example when selecting metadata columns.  

It's a bit verbose to me, I dont know what you guys think, if we should remove logging the exception stack?  

```
[Test worker] WARN org.apache.iceberg.spark.source.SparkScanBuilder - Failed to bind expression to table schema, skipping push down for this expression: EqualTo(_spec_id,0)
org.apache.iceberg.exceptions.ValidationException: Cannot find field '_spec_id' in struct: struct<2147483546: file_path: required string (Path of a file in which a deleted row is stored), 2147483545: pos: required long (Ordinal position of a deleted row in the data file), 2147483544: row: optional struct<1: id: required int, 2: data: required string> (Deleted row values), 2147483543: partition: required struct<1000: data: optional string, 1001: id: optional int> (Partition to which a row belongs to)>
	at org.apache.iceberg.exceptions.ValidationException.check(ValidationException.java:50)
	at org.apache.iceberg.expressions.NamedReference.bind(NamedReference.java:46)
	at org.apache.iceberg.expressions.NamedReference.bind(NamedReference.java:27)
	at org.apache.iceberg.expressions.UnboundPredicate.bind(UnboundPredicate.java:106)
	at org.apache.iceberg.expressions.Binder$BindVisitor.predicate(Binder.java:164)
	at org.apache.iceberg.expressions.Binder$BindVisitor.predicate(Binder.java:123)
	at org.apache.iceberg.expressions.ExpressionVisitors.visit(ExpressionVisitors.java:330)
	at org.apache.iceberg.expressions.Binder.bind(Binder.java:63)
	at org.apache.iceberg.spark.source.SparkScanBuilder.pushFilters(SparkScanBuilder.java:117)
	at org.apache.spark.sql.execution.datasources.v2.PushDownUtils$.pushFilters(PushDownUtils.scala:65)
	at org.apache.spark.sql.execution.datasources.v2.V2ScanRelationPushDown$$anonfun$pushDownFilters$1.applyOrElse(V2ScanRelationPushDown.scala:72)
	at org.apache.spark.sql.execution.datasources.v2.V2ScanRelationPushDown$$anonfun$pushDownFilters$1.applyOrElse(V2ScanRelationPushDown.scala:59)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$1(TreeNode.scala:584)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:176)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning(TreeNode.scala:584)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.org$apache$spark$sql$catalyst$plans$logical$AnalysisHelper$$super$transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning(AnalysisHelper.scala:267)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning$(AnalysisHelper.scala:263)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$3(TreeNode.scala:589)
	at org.apache.spark.sql.catalyst.trees.UnaryLike.mapChildren(TreeNode.scala:1228)
	at org.apache.spark.sql.catalyst.trees.UnaryLike.mapChildren$(TreeNode.scala:1227)
	at org.apache.spark.sql.catalyst.plans.logical.OrderPreservingUnaryNode.mapChildren(LogicalPlan.scala:208)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning(TreeNode.scala:589)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.org$apache$spark$sql$catalyst$plans$logical$AnalysisHelper$$super$transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning(AnalysisHelper.scala:267)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning$(AnalysisHelper.scala:263)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:560)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transform(TreeNode.scala:528)
	at org.apache.spark.sql.execution.datasources.v2.V2ScanRelationPushDown$.pushDownFilters(V2ScanRelationPushDown.scala:59)
	at org.apache.spark.sql.execution.datasources.v2.V2ScanRelationPushDown$.$anonfun$apply$3(V2ScanRelationPushDown.scala:44)
	at org.apache.spark.sql.execution.datasources.v2.V2ScanRelationPushDown$.$anonfun$apply$7(V2ScanRelationPushDown.scala:50)
	at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126)
	at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122)
	at scala.collection.immutable.List.foldLeft(List.scala:91)
	at org.apache.spark.sql.execution.datasources.v2.V2ScanRelationPushDown$.apply(V2ScanRelationPushDown.scala:49)
	at org.apache.spark.sql.execution.datasources.v2.V2ScanRelationPushDown$.apply(V2ScanRelationPushDown.scala:37)
	at org.apache.spark.sql.catalyst.rules.RuleExecutor.$anonfun$execute$2(RuleExecutor.scala:211)
	at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126)
	at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122)
	at scala.collection.immutable.List.foldLeft(List.scala:91)
	at org.apache.spark.sql.catalyst.rules.RuleExecutor.$anonfun$execute$1(RuleExecutor.scala:208)
	at org.apache.spark.sql.catalyst.rules.RuleExecutor.$anonfun$execute$1$adapted(RuleExecutor.scala:200)
	at scala.collection.immutable.List.foreach(List.scala:431)
	at org.apache.spark.sql.catalyst.rules.RuleExecutor.execute(RuleExecutor.scala:200)
	at org.apache.spark.sql.catalyst.rules.RuleExecutor.$anonfun$executeAndTrack$1(RuleExecutor.scala:179)
	at org.apache.spark.sql.catalyst.QueryPlanningTracker$.withTracker(QueryPlanningTracker.scala:88)
	at org.apache.spark.sql.catalyst.rules.RuleExecutor.executeAndTrack(RuleExecutor.scala:179)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$optimizedPlan$1(QueryExecution.scala:126)
	at org.apache.spark.sql.catalyst.QueryPlanningTracker.measurePhase(QueryPlanningTracker.scala:111)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$executePhase$2(QueryExecution.scala:185)
	at org.apache.spark.sql.execution.QueryExecution$.withInternalError(QueryExecution.scala:510)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$executePhase$1(QueryExecution.scala:185)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.execution.QueryExecution.executePhase(QueryExecution.scala:184)
	at org.apache.spark.sql.execution.QueryExecution.optimizedPlan$lzycompute(QueryExecution.scala:122)
	at org.apache.spark.sql.execution.QueryExecution.optimizedPlan(QueryExecution.scala:118)
	at org.apache.spark.sql.execution.QueryExecution.assertOptimized(QueryExecution.scala:136)
	at org.apache.spark.sql.execution.QueryExecution.executedPlan$lzycompute(QueryExecution.scala:154)
	at org.apache.spark.sql.execution.QueryExecution.executedPlan(QueryExecution.scala:151)
	at org.apache.spark.sql.execution.QueryExecution.simpleString(QueryExecution.scala:204)
	at org.apache.spark.sql.execution.QueryExecution.org$apache$spark$sql$execution$QueryExecution$$explainString(QueryExecution.scala:249)
	at org.apache.spark.sql.execution.QueryExecution.explainString(QueryExecution.scala:218)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:103)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:169)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:95)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3856)
	at org.apache.spark.sql.Dataset.collectAsList(Dataset.scala:3131)
	at org.apache.iceberg.spark.source.TestPositionDeletesTable.actualPositionDeleteRowSet(TestPositionDeletesTable.java:881)
	at org.apache.iceberg.spark.source.TestPositionDeletesTable.testPartitionEvolutionSpecId(TestPositionDeletesTable.java:729)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$2.run(TestWorker.java:176)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
[Test worker] INFO org.apache.spark.sql.execution.datasources.v2.V2ScanRelationPushDown - 

```

**Comment by bryanck** [comment]
That makes sense to me, also we should probably lower the log level to `DEBUG` if this is something that happens under normal circumstances

**Comment by huaxingao** [comment]
I will remove the exception stack. I can also lower the log level to `DEBUG`.

**Comment by szehon-ho** [comment]
OK, I'm fine with INFO, i think it's if the user selects a column without stats like struct or metadata column, I'm betting it shouldn't  be every query, but we can change it if it is too much.  Thanks guys !

**Comment by kbendick** [comment]
> OK, I'm fine with INFO, i think it's if the user selects a column without stats like struct or metadata column, I'm betting it shouldn't be every query, but we can change it if it is too much. I think we can keep logging the exception toString too, just not the whole stack. Thanks guys !

I think INFO, but without the whole stack trace (just the toString or the message), is best.

Not having a filter pushed down when you expect it to be pushed down can result in scanning huge amounts of data, so relative to some of the INFO level logs that come out of Spark, so I think it’s fine even as WARN (though if we expect it often enough with metadata columns than INFO is probably better).

So +1 to dealing with the big stack trace but keeping the message. Good catch @szehon-ho!

**Pull request merged**